### PR TITLE
feat(runtime): support for native tool calling

### PR DESCRIPTION
**Pull request by mhamann**
### **PR Type**
Enhancement

___

### **Description**
- Add native tool calling support with XML-based tool call detection

- Implement prompt-based tool injection for models without native support

- Support Jinja2 template rendering for tool-aware chat templates

- Handle tool calls in streaming and non-streaming chat completions

- Add tool call detection and extraction from model responses

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Chat Request<br/>with Tools"] --> B["Prepare Messages<br/>with Tools"]
  B --> C{Model has<br/>Tool Template?}
  C -->|Yes| D["Jinja2 Render<br/>with Tools"]
  C -->|No| E["Inject Tools<br/>into Prompt"]
  D --> F["Generate Response"]
  E --> F
  F --> G["Detect Tool Calls<br/>in Content"]
  G --> H{Tool Calls<br/>Found?}
  H -->|Yes| I["Return Tool Calls<br/>in Response"]
  H -->|No| J["Return Text<br/>Response"]
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>llama.py</strong><dd><code>Handle None content in chat messages</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/636/files#diff-cb2f5fd3e3544c1abb28deac6e74d566628635b5dd4d9bdb64b84f45f2d70559">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Enhancement</strong></td><td><details><summary>5 files</summary><table>
<tr>
  <td><strong>gguf_language_model.py</strong><dd><code>Add tool calling support to language model</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/636/files#diff-aa3d47952f93f3f6d338873bbc86b4f9eb2700b79e17fd3e73e5a2016fc37064">+66/-2</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>service.py</strong><dd><code>Implement tool call detection in chat completions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/636/files#diff-0d5bb320d0590e4a8fb439fd35b07be371bae7d283b4862979f25fbaad05be14">+142/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>types.py</strong><dd><code>Add tool calling request types and models</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/636/files#diff-03652a29495c88457d7e5e4bad3f5e17fdab9f7a9a2da52720da0302ca0cddd2">+20/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>tool_calling.py</strong><dd><code>Implement prompt-based tool calling utilities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/636/files#diff-444c620588ed78f7e545039372f306e1b3a574937a9df569f673253b87a9c1ea">+154/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>jinja_tools.py</strong><dd><code>Add Jinja2 template rendering for tools</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/636/files#diff-2c60e19ce7029ccfcc121a813eb934a55373ec307a87a7771de99cc631396855">+176/-0</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>test_tool_calling.py</strong><dd><code>Add comprehensive tests for tool calling</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/636/files#diff-148967cb2010c7d664ee8e0e680bcdae82346e598aaf98e9c2fc6815314729f8">+198/-0</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Dependencies</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>pyproject.toml</strong><dd><code>Add jinja2 dependency for template rendering</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/636/files#diff-4784ecbf26f867a617e29d2d9c3543bad45bc93f335b7b9403e603218265152c">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>project.json</strong><dd><code>Add nx project configuration file</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/636/files#diff-d2d4562952d468fe80f55c81244ec427178e9b737fd8e200ae10714b7147adf7">+76/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tbody></table>

</details>

___

### CODE REVIEW

**Comment by BobbyRadford** [comment]
## Code Review Summary

Overall this is a solid implementation of native tool calling support. The architecture with Jinja2 templates + prompt-based fallback is well thought out, and the test coverage is good.

### Key Findings

| Severity | Issue | Location |
|----------|-------|----------|
| 🟡 Medium | Misleading log message - says "using Jinja2 rendering" but always falls through to prompt injection | `gguf_language_model.py:299-312` |
| 🟡 Medium | `tool_choice` parameter accepted but not implemented | `gguf_language_model.py:289-293` |
| 🔵 Low | Scope creep - context management (truncation/summarization) bundled with tool calling | `service.py` |
| 🔵 Low | Only first tool call detected if model outputs multiple | `tool_calling.py` |

### Questions

1. Is the context management intentionally bundled here, or should it be a separate PR?
2. What's the timeline for implementing the Jinja2 native tool template path?
3. Should `tool_choice` enforcement be added before merge?

### Test Coverage Gaps

- Streaming with tool calls
- Multiple tool calls in single response  
- Edge cases (malformed XML, partial tool calls)

**Verdict: Approve with suggestions** - Address the misleading Jinja2 log message before merge. Other items are non-blocking.

File: runtimes/universal/models/gguf_language_model.py [hunk 1]
```diff
@@ -219,6 +233,106 @@ def format_messages(self, messages: list[dict]) -> str:
...
+                f"tool_choice='{tool_choice}' was specified but is not yet implemented. "
+                "The model will decide whether to use tools based on the prompt."
+            )
+
+        # Try Jinja2 rendering for models with tool-aware templates
+        try:
+            from utils.jinja_tools import (
+                get_chat_template_from_gguf,
+                supports_native_tools,
+            )
```
Comment (BobbyRadford) [inline runtimes/universal/models/gguf_language_model.py]: **Misleading log message**: This logs "using Jinja2 rendering" but then immediately falls through to prompt-based injection. The Jinja2 path is never actually used.

Suggestion:
```python
logger.debug(
    "Model has tool-aware template but falling back to prompt injection "
    "(Jinja2 rendering not yet fully implemented)"
)
```

Or implement the Jinja2 path before the TODO comment.

File: runtimes/universal/models/gguf_language_model.py [hunk 2]
```diff
@@ -219,6 +233,106 @@ def format_messages(self, messages: list[dict]) -> str:
...
+        """Prepare messages with tool definitions if provided.
+
+        Tries Jinja2 template rendering first for models with native tool support,
+        falls back to prompt-based injection.
+
+        Args:
+            messages: List of message dicts with 'role' and 'content' keys
+            tools: Optional list of tool definitions in OpenAI format
+            tool_choice: Tool choice strategy (not yet implemented, accepted for API compatibility)
+
```
Comment (BobbyRadford) [inline runtimes/universal/models/gguf_language_model.py:415]: Consider whether accepting `tool_choice` without implementing it is the right UX. Options:
1. Return HTTP 501 Not Implemented
2. Include in response metadata that it was ignored
3. At minimum, document this limitation in API docs

The warning log is good but clients may not see it.

File: runtimes/universal/routers/chat_completions/service.py [hunk 3]
```diff
@@ -110,6 +117,151 @@ async def chat_completions(self, chat_request: ChatCompletionRequest):
...
+            tools_dict = None
+            if chat_request.tools:
+                tools_dict = [dict(tool) for tool in chat_request.tools]
+
+            # Context management for GGUF models
+            context_usage_info = None
+            if is_gguf and model.context_manager:
+                # Apply history compression to reduce token usage
+                compressor = HistoryCompressor(model.token_counter)
+                messages_dict = compressor.compress(messages_dict)
```
Comment (BobbyRadford) [inline runtimes/universal/routers/chat_completions/service.py:148]: The context management code adds significant complexity to this PR. Consider splitting into a separate PR for:
1. Easier review
2. Cleaner git history
3. Independent rollback if issues arise

This is a suggestion, not a blocker.

**Comment by BobbyRadford** [comment]
### Additional Note on `tool_calling.py`

**Multiple tool calls handling**: The `detect_tool_call_in_content` function only returns the first match. If a model outputs multiple tool calls (e.g., parallel function calling like OpenAI supports), subsequent calls would be lost.

Is this intentional? If so, please document this limitation. If not, consider:
```python
def detect_tool_calls_in_content(content: str) -> list[dict]:
    """Detect ALL tool calls in content, not just the first."""
    ...
```

File: server/agents/base/clients/openai.py [hunk 4]
```diff
@@ -83,7 +83,15 @@ async def chat(
         )
 
         # Convert tools to OpenAI format
-        if self._model_config.tool_call_strategy == ToolCallStrategy.native_api:
+        # Check for native_api strategy (handle both enum and string values)
```
Comment (BobbyRadford) [inline server/agents/base/clients/openai.py:86]: Seems a bit weird that we could have enum and string values

File: runtimes/universal/routers/chat_completions/router.py [hunk 5]
```diff
@@ -13,4 +16,11 @@ async def chat_completions(chat_request: ChatCompletionRequest):
 
     Supports any HuggingFace causal language model.
     """
+    # Debug log the incoming request
+    if logger.isEnabledFor(logging.DEBUG):
```
Comment (BobbyRadford) [inline runtimes/universal/routers/chat_completions/router.py:20]: Why do we check logging level instead of just logging and letting the logger sort that out?

Comment (mhamann) [reply runtimes/universal/routers/chat_completions/router.py:20]: great question

File: server/agents/base/clients/openai.py (hunk 4 above)
Comment (mhamann) [reply server/agents/base/clients/openai.py:86]: double checking this

Comment (mhamann) [reply server/agents/base/clients/openai.py:86]: The generated datamodel.py file seems to use a string as its default value, even though an enum should be used. Unless it gets coerced later, then we'd still need to check it.

File: runtimes/universal/routers/chat_completions/router.py (hunk 5 above)
Comment (mhamann) [reply runtimes/universal/routers/chat_completions/router.py:20]: ah, these were added during performance tuning. calls to functions like `json.dumps()` can be expensive, so this prevents that from occurring if debug logging isn't enabled
